### PR TITLE
Preliminary work to add Playwright E2E tests for the Cart block & merchant cart tests

### PR DESCRIPTION
**Pull request by opr**
This PR lays some groundwork for adding Cart block tests using Playwright and adds some 'merchant' tests for it. There are still more tests to convert, but this is a good starting point.

The following changes are included:

- Adds the navigation.ts file which contains some navigation-related utils (closeModalIfExists and editBlockPage) respectively, these will close any modals on the editor page, and open the site editor for a specific block page, for example if you pass "Checkout Block" to the editBlockPage function, playwright will open the site editor for the page called "Checkout Block".
- Augments the `EditorUtils` class to include a reference to the `page` and adds three new functions which aren't available through the Gutenberg e2e utils package
  - `isGlobalInserterOpen`,
  - `openGlobalBlockInserter`,
  - `toggleGlobalBlockInserter`

Supersedes woocommerce/woocommerce-blocks#9062

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### Developer Testing

1. Run `npm i`
2. `npm run env:start` 
3. `npm run test:e2e-pw`
4. Ensure all tests pass.
5. Visit the actions tab on this PR and view the E2E Playwright tests (introduced in woocommerce/woocommerce-blocks#9034) and ensure they pass.

* [x] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add suggested changelog entry here.

### CODE REVIEW

File: tests/e2e-pw/playwright.config.ts [hunk 1]
```diff
@@ -67,6 +67,9 @@ const config: ExtendedPlaywrightTestConfig = {
 			testMatch: /.*.classic_theme.spec.ts/,
 			dependencies: [ 'classicThemeConfiguration' ],
 		},
+		{
+			name: 'cartCheckout',
+		},
```
Comment (nielslange) [inline tests/e2e-pw/playwright.config.ts]: For other entries, I see that they have `testMatch` and `dependencies`. Is this not needed in this case?

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+					'{ additionalCartCheckoutInnerBlockTypes: ( value, extensions, { block } ) => {' +
+					"    value.push('core/table');" +
+					"    if ( block === 'woocommerce/cart-order-summary-block' ) {" +
+					"        value.push( 'core/audio' );" +
+					'    }' +
+					'    return value;' +
+					'}' +
+					'}' +
+					');'
+			);
```
Comment (nielslange) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: Nice one, @opr! 👏

File: tests/e2e-pw/utils/editor/EditorUtils.ts [hunk 3]
```diff
@@ -2,17 +2,71 @@
...
+		} );
+	}
+
+	/**
+	 * Opens the global inserter.
+	 */
+	async openGlobalBlockInserter() {
+		if ( ! ( await this.isGlobalInserterOpen() ) ) {
+			await this.toggleGlobalBlockInserter(); // Waiting here is necessary because sometimes the inserter takes more
+			// time to render than Puppeteer takes to complete the 'click' action.
```
Comment (nielslange) [inline tests/e2e-pw/utils/editor/EditorUtils.ts]: What are your thoughts on placing the comment above the function call?

File: tests/e2e-pw/utils/navigation/navigation.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Closes any modals in the editor if they are open.
+ */
+export const closeModalIfExists = async (
+	page: PlaywrightTestArgs[ 'page' ]
+) => {
+	// The modal close button can have different aria-labels, depending on the version of Gutenberg/WP.
+	// Newer versions (WP >=6.2) use `Close`, while older versions (WP <6.1) use `Close dialog`.
+	const closeButton = await page.$(
+		'.components-modal__header [aria-label="Close"], .components-modal__header [aria-label="Close dialog"]'
```
Comment (nielslange) [inline tests/e2e-pw/utils/navigation/navigation.ts]: I wonder if we can use 
```css
'.components-modal__header [aria-label^="Close"]'
```
or
```css
'.components-modal__header [aria-label*="Close"]'
```
instead of 
```css
'.components-modal__header [aria-label="Close"], .components-modal__header [aria-label="Close dialog"]'
```

https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors states:
> [attr^=value]
> Represents elements with an attribute name of attr whose value is prefixed (preceded) by value.
>
> [attr*=value]
> Represents elements with an attribute name of attr whose value contains at least one occurrence of value within the string.

`[attr^=value]` can be used to check is a string starts with something, while `[attr*=value]` can be used to check if the string contains something. In this particular case, I'd say that `[attr^=value]` is more suitable. What are your thoughts on this?

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @opr. While reviewing this PR, I noticed that it contains a merge conflict due to https://github.com/woocommerce/woocommerce-blocks/pull/9951. Apart from that, I noticed that the Playwright tests timed out after 60 min. 

Besides these two things, I left a few comments.

File: tests/e2e-pw/utils/editor/EditorUtils.ts (hunk 3 above)
Comment (opr) [reply tests/e2e-pw/utils/editor/EditorUtils.ts]: Removing the comment altogether is better I think, we don't wait anymore.

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts (hunk 2 above)
Comment (opr) [reply tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: Thanks @nielslange

File: tests/e2e-pw/playwright.config.ts (hunk 1 above)
Comment (opr) [reply tests/e2e-pw/playwright.config.ts]: Good point, I added it so I could control better which tests I ran in isolation, but it's a nicer idea to include a pattern to test on, you're right.

File: tests/e2e-pw/utils/navigation/navigation.ts (hunk 4 above)
Comment (opr) [reply tests/e2e-pw/utils/navigation/navigation.ts]: Thanks for the suggestion and for writing up the ^ and * selectors.

In this case I think I would prefer to leave it as is. In my opinion it is better to be explicit about what we're looking for instead of using wildcards.

Comment (nielslange) [reply tests/e2e-pw/utils/navigation/navigation.ts]: Being explicit sounds good, @opr. Just wanted to point out an alternative. That said, with your comment, I think being explicit is indeed the better approach here.

**Review by nielslange — COMMENTED** [review]
I noticed that in some cases running `npm run env:start` lead to an error that the `wc` command is unknown. What are your thoughts on mentioning `npm run env:restart` in the testing steps as an alternative if `npm run env:start` is leading to issues?

Looking at the CI pipeline, I noticed that some Playwright tests are still failing. I triggered the CI pipeline again to see if the failing tests were flaky tests.

I also noticed that we're currently only using 1 worker. The corresponding part can be found in https://github.com/woocommerce/woocommerce-blocks/blob/1b9746e168037291f4343eb4d801b49822e4b22f/tests/e2e-pw/playwright.config.ts#L30

I have the impression that by having 1 worker, the tests are running rather long. Locally, I increased the number of workers to 10 and the tests finished much faster. Regarding worker, https://playwright.dev/docs/test-parallel states the following:
> _"Playwright Test runs tests in parallel. In order to achieve that, it runs several worker processes that run at the same time. By default, test files are run in parallel."_

I wonder if we should increase the number of worker to decrease the total running time of the tests. @gigitux it would be great to hear your thoughts on this as well, as you lowered the number of workers from 4 to 1 in #9551.

**Comment by gigitux** [comment]
> I wonder if we should increase the number of worker to decrease the total running time of the tests. @gigitux it would be great to hear your thoughts on this as well, as you lowered the number of workers from 4 to 1 in woocommerce/woocommerce-blocks#9551.

Yes, we should increase the number of workers, but currently, we have a blocker:
https://github.com/woocommerce/woocommerce-blocks/issues/9632

The E2E tests with side effects (so with the `.side_effects suffix`) should continue running one at a time. You can find more details in the issue that I shared.

**Comment by nielslange** [comment]
> Yes, we should increase the number of workers, but currently, we have a blocker: woocommerce/woocommerce-blocks#9632
> 
> The E2E tests with side effects (so with the `.side_effects suffix`) should continue running one at a time. You can find more details in the issue that I shared.

Thanks for explaining why the number of worker has been changed from 4 to 1, @gigitux. 

@gigitux and @opr While running `npm run test:e2e-pw -- tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts`, I saw the following message:
> Slow test file: [cartCheckout] › cart-block/cart-block.merchant.spec.ts (16.6s)
> Consider splitting slow test files to speed up parallel execution

Based on that message, I split the file `cart-block.merchant.spec.ts` into the following files (thus, one file for each test case):
- `cart-block.merchant-can-be-inserted-once.spec.ts`
- `cart-block.merchant-can-be-removed-by-filter.spec.ts`
- `cart-block.merchant-shows-empty-cart.spec.ts`
- `cart-block.merchantrender-without-crashing.spec.ts`

Screenshot 2023-06-29 at 11 55 16

This change decreased the running time from 17.0s to 11.2s, meaning a speed improvement by 34%. That said, I did increase the number of workers when running this experiment. 

What are your thoughts on having one use case per file, once we addressed woocommerce/woocommerce-blocks#9632? Based on my understanding, it appears to be better to have a lot of tiny test files than a handful of large test files.

**Comment by gigitux** [comment]
> > Yes, we should increase the number of workers, but currently, we have a blocker: woocommerce/woocommerce-blocks#9632
> > The E2E tests with side effects (so with the `.side_effects suffix`) should continue running one at a time. You can find more details in the issue that I shared.
> 
> Thanks for explaining why the number of worker has been changed from 4 to 1, @gigitux.
> 
> @gigitux and @opr While running `npm run test:e2e-pw -- tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts`, I saw the following message:
> 
> > Slow test file: [cartCheckout] › cart-block/cart-block.merchant.spec.ts (16.6s)
> > Consider splitting slow test files to speed up parallel execution
> 
> Based on that message, I split the file `cart-block.merchant.spec.ts` into the following files (thus, one file for each test case):
> 
> * `cart-block.merchant-can-be-inserted-once.spec.ts`
> * `cart-block.merchant-can-be-removed-by-filter.spec.ts`
> * `cart-block.merchant-shows-empty-cart.spec.ts`
> * `cart-block.merchantrender-without-crashing.spec.ts`
> 
> This change decreased the running time from 17.0s to 11.2s, meaning a speed improvement by 34%. What are your thoughts on having one use case per file, once we addressed woocommerce/woocommerce-blocks#9632? Based on my understanding, it appears to be better to have a lot of tiny test files than a handful of large test files.

We should adopt this approach when the E2E tests are very long (like 1.5 minutes). Also, when woocommerce/woocommerce-blocks#9362 will be fixed, those E2E tests will go very fast.

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts
Comment (gigitux) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts:1]: Mmm, good observation adding the `merchant` suffix. Do you think we should have two suffixes: one for E2E tests on the frontend side and another for the editor side? 

Also, be careful that currently missing the suffix to set on witch configuration this test should run ([details](https://github.com/woocommerce/woocommerce-blocks/blob/fa3f9856629dcbf77d12fc9064b34b9820574776/docs/contributors/e2e-guidelines.md/#L19-L25))

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
+/**
+ * External dependencies
+ */
+import { BlockData } from '@woocommerce/e2e-types';
+import { test, expect } from '@woocommerce/e2e-playwright-utils';
+
+/**
+ * Internal dependencies
+ */
+import { editBlockPage } from '../../utils/navigation/navigation';
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: To make the codebase cleaner, we should import the [function from the packages](https://github.com/woocommerce/woocommerce-blocks/blob/fa3f9856629dcbf77d12fc9064b34b9820574776/docs/contributors/e2e-guidelines.md/#L51-L52).

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+			insertButton: "//button//span[text()='Cart']",
+		},
+		frontend: {
+			block: '.wp-block-woocommerce-cart',
+		},
+	},
+};
+
+test.describe( 'Merchant → Cart', () => {
+	test.use( { storageState: process.env.ADMINSTATE } );
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: Is it necessary?

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+				"wc.blocksCheckout.registerCheckoutFilters( 'woo-test-namespace'," +
+					'{ additionalCartCheckoutInnerBlockTypes: ( value, extensions, { block } ) => {' +
+					"    value.push('core/table');" +
+					"    if ( block === 'woocommerce/cart-order-summary-block' ) {" +
+					"        value.push( 'core/audio' );" +
+					'    }' +
+					'    return value;' +
+					'}' +
+					'}' +
+					');'
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: NIT: If you prefer, you can use the ` to avoid string concatenation:

```suggestion
				`wc.blocksCheckout.registerCheckoutFilters( 'woo-test-namespace', {
	additionalCartCheckoutInnerBlockTypes: ( value, extensions, { block } ) => {
		value.push( 'core/table' );
		if ( block === 'woocommerce/cart-order-summary-block' ) {
			value.push( 'core/audio' );
		}
		return value;
	},
} );`
```

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+			// Focus the empty cart button and wait for the tooltip to disappear.
+			await emptyCartButton.focus();
+			await page.waitForTimeout( 1000 );
+			await emptyCartButton.click();
+
+			await expect(
+				page.locator(
+					blockData.selectors.editor.block +
+						' [data-type="woocommerce/filled-cart-block"]'
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: Is there a reason why you didn't use [`getBlockByName`](https://github.com/woocommerce/woocommerce-blocks/blob/fa3f9856629dcbf77d12fc9064b34b9820574776/tests/e2e-pw/utils/editor/EditorUtils.ts/#L19-L21)?

File: tests/e2e-pw/types/block-data.ts [hunk 9]
```diff
@@ -1,5 +1,6 @@
 export type BlockData< T = unknown > = {
 	name: string;
+	slug: string;
 	mainClass: string;
-	selectors: Record< 'editor' | 'frontend', Record< string, unknown > >;
+	selectors: Record< 'editor' | 'frontend', Record< string, string > >;
```
Comment (gigitux) [inline tests/e2e-pw/types/block-data.ts]: Not sure about this change. It could be necessary having an object like this:

```
	selectors: {
		frontend: {
			header: '.blabla',
			footer: '.blablabla',
		},
	},
```

File: tests/e2e-pw/utils/navigation/navigation.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Closes any modals in the editor if they are open.
+ */
+export const closeModalIfExists = async (
+	page: PlaywrightTestArgs[ 'page' ]
+) => {
+	// The modal close button can have different aria-labels, depending on the version of Gutenberg/WP.
+	// Newer versions (WP >=6.2) use `Close`, while older versions (WP <6.1) use `Close dialog`.
+	const closeButton = await page.$(
```
Comment (gigitux) [inline tests/e2e-pw/utils/navigation/navigation.ts]: It always better to use `[locator](https://github.com/woocommerce/woocommerce-blocks/blob/fa3f9856629dcbf77d12fc9064b34b9820574776/docs/contributors/e2e-guidelines.md/#L45-L46)`

File: tests/e2e-pw/utils/navigation/navigation.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+	if ( closeButton ) {
+		await closeButton.click();
+	}
+};
+
+/**
+ * Goes to the edit page of a specified block.
+ */
+export const editBlockPage = async (
+	page: PlaywrightTestArgs[ 'page' ],
```
Comment (gigitux) [inline tests/e2e-pw/utils/navigation/navigation.ts:23]: You can use `Page` instead of `PlaywrightTestArgs[ 'page' ]`.

File: tests/e2e-pw/utils/navigation/navigation.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import { PlaywrightTestArgs } from '@playwright/test';
+import { BlockData } from '@woocommerce/e2e-types';
+
+/**
+ * Closes any modals in the editor if they are open.
+ */
+export const closeModalIfExists = async (
+	page: PlaywrightTestArgs[ 'page' ]
+) => {
+	// The modal close button can have different aria-labels, depending on the version of Gutenberg/WP.
```
Comment (gigitux) [inline tests/e2e-pw/utils/navigation/navigation.ts:11]: Good approach. I wonder how it could be a more scalable approach 🤔 . I will think about it when I will work on https://github.com/woocommerce/woocommerce-blocks/issues/9003

File: tests/e2e-pw/playwright.config.ts [hunk 13]
```diff
@@ -67,6 +67,10 @@ const config: ExtendedPlaywrightTestConfig = {
 			testMatch: /.*.classic_theme.spec.ts/,
 			dependencies: [ 'classicThemeConfiguration' ],
 		},
+		{
+			name: 'cartCheckout',
+			testMatch: /cart-block.*/,
+		},
```
Comment (gigitux) [inline tests/e2e-pw/playwright.config.ts]: Why did you add a new configuration?

**Review by gigitux — CHANGES_REQUESTED** [review]
Nice work! 
I added some feedback! Let me know what you think about!

File: tests/e2e-pw/playwright.config.ts (hunk 13 above)
Comment (opr) [reply tests/e2e-pw/playwright.config.ts]: If we want to split out the FSE tests and Cart/Checkout tests (I did not add the `checkout-block` directory because it doesn't exist yet) this is the best way to achieve that?

Comment (gigitux) [reply tests/e2e-pw/playwright.config.ts]: Do you think we should do it?

Comment (opr) [reply tests/e2e-pw/playwright.config.ts]: I think so, during the Woo DM we had a team time meeting, and we spoke about splitting the cart/checkout tests and the FSE tests, most of the team were in favour of that, cc @ralucaStan do you still support this?

File: tests/e2e-pw/types/block-data.ts (hunk 9 above)
Comment (opr) [reply tests/e2e-pw/types/block-data.ts]: We should try to define this more robustly, I'll revert this for now!

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts (hunk 8 above)
Comment (opr) [reply tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: No, I have updated it now! Thanks for the reminder about that util :)

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts (hunk 5 above)
Comment (opr) [reply tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts]: This is a local util, it's not part of that package. Do you think we should make an alias for local packages?

**Comment by gigitux** [comment]
@opr I noticed that you updated the PR. Can I review the PR?

**Comment by nielslange** [comment]
> @opr I noticed that you updated the PR. Can I review the PR?

@gigitux Our team meetup in Antalya starts today so @opr might be travelling right now and therefore reply a bit later than usual.

**Comment by gigitux** [comment]
> > @opr I noticed that you updated the PR. Can I review the PR?
> 
> @gigitux Our team meetup in Antalya starts today so @opr might be travelling right now and therefore reply a bit later than usual.

No worries! It is not urgent! Enjoy the team meetup!

File: tests/e2e-pw/playwright.config.ts (hunk 13 above)
Comment (alexflorisca) [reply tests/e2e-pw/playwright.config.ts]: I think this is valuable also, as we are 2 different teams, we may want to run just cart & checkout tests and you guys might want to run the FSE tests. If not in CI, then definitely locally for speed of development. What might get tricky here is the permutations. For example, I notice there are `blockTheme` and `classicTheme` projects. What happens if we want to test the cart checkout for both classic and block themes? Would the tests go in the `cartCheckout` project or one of the other two mentioned above?

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts
Comment (alexflorisca) [reply tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts:1]: Can I suggest a file structure of `tests/cart-checkout/cart-block.merchant.spec.ts` instead? The reason being these are e2e tests, and some of them will cross over between cart and checkout pages to test the whole flow. It makes sense to keep them in the same place rather then split them by blocks in this context.

**Comment by opr** [comment]
@gigitux rebased, and fixed some issues with the cart tests following our move to templates. Please review when you're able to! 🙏🏼

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.block_theme.spec.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
+/**
```
Comment (alexflorisca) [inline tests/e2e-pw/tests/cart-block/cart-block.merchant.block_theme.spec.ts]: I left a comment earlier but I think it got lost once you pushed some changes. Anyway I was suggesting a file structure of `tests/cart-checkout/testname.spec.ts`. Mainly because we're dealing with e2e tests here, some of which will span both cart and checkout blocks. So it makes sense to group them together here rather than separate directories by blocks.

File: tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts
Comment (opr) [reply tests/e2e-pw/tests/cart-block/cart-block.merchant.spec.ts:1]: Yeah, agree with you there actually, I updated the folder name!

File: tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+					},
+				} );`
+			);
+
+			await editor.insertBlock( { name: 'woocommerce/cart' } );
+			await expect(
+				await editorUtils.getBlockByName( blockData.slug )
+			).not.toHaveCount( 0 );
+
+			// // Select the cart-order-summary-block block and try to insert a block. Check the Table block is available.
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts]: NIT: Double comment

```suggestion
			// Select the cart-order-summary-block block and try to insert a block. Check the Table block is available.
```

File: tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+					blockData.selectors.editor.block +
+						' [data-type="woocommerce/filled-cart-block"]'
+				)
+			).toBeVisible();
+			await expect(
+				editor.canvas.locator(
+					blockData.selectors.editor.block +
+						' [data-type="woocommerce/empty-cart-block"]'
+				)
+			).not.toBeVisible();
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts]: We can refactor this to `.toBeHidden`. I'm adding `eslint-plugin-playwright` package that has [this rule](https://github.com/playwright-community/eslint-plugin-playwright/blob/main/docs/rules/no-useless-not.md).

```suggestion
			).toBeHidden();
```

File: tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+			insertButton: "//button//span[text()='Cart']",
+		},
+		frontend: {
+			block: '.wp-block-woocommerce-cart',
+		},
+	},
+};
+
+test.describe( 'Merchant → Cart', () => {
+	const blockSelectorInEditor = blockData.selectors.editor.block as string;
```
Comment (gigitux) [inline tests/e2e-pw/tests/cart-checkout/cart-block.merchant.block_theme.spec.ts]: Is `as string` necessary?

**Review by gigitux — APPROVED** [review]
Great work! I added some comments but I will pre-approve this PR 🚢
